### PR TITLE
feat: better handle inheritance

### PR DESCRIPTION
**Pull request by pollend**
- include super class properties to inject into
- don't instantiate interfaces/abstract classes

### CODE REVIEW

File: gestalt-inject-java/src/main/java/org/terasology/gestalt/annotation/processing/BeanDefinitionProcessor.java [hunk 1]
```diff
@@ -304,23 +310,34 @@ private CodeBlock buildResolution(String method, VariableElement element, ClassN
...
         private CodeBlock buildInjectionBlock(TypeElement target) {
             String name = "instance";
             CodeBlock.Builder builder = CodeBlock.builder();
 
-            for (Element ele : target.getEnclosedElements()) {
-                if (ele.getAnnotationMirrors().stream().anyMatch(in -> in.getAnnotationType().toString().equals(Inject.class.getName()))) {
-                    if (ele instanceof ExecutableElement) {
-                        List<? extends VariableElement> parameters = ((ExecutableElement) ele).getParameters();
+            TypeElement targetElement = target;
+            while (true) {
```
Comment (DarkWeird) [inline gestalt-inject-java/src/main/java/org/terasology/gestalt/annotation/processing/BeanDefinitionProcessor.java]: Don't use `while (true)` please.
I see how this works... but this line terrifies me.

File: gestalt-inject-java/src/main/java/org/terasology/gestalt/annotation/processing/BeanDefinitionProcessor.java [hunk 2]
```diff
@@ -264,6 +265,11 @@ private CodeBlock buildObjectPointCreationBlock(TypeElement target) {
             CodeBlock.Builder builder = CodeBlock.builder();
             ClassName targetClass = ClassName.get(target);
 
+            // abstract/interface classes can't be instantiated
+            if(target.getModifiers().contains(Modifier.ABSTRACT) || target.getKind() == ElementKind.INTERFACE) {
+                return builder.add("return Optional.empty();").build();
```
Comment (DarkWeird) [inline gestalt-inject-java/src/main/java/org/terasology/gestalt/annotation/processing/BeanDefinitionProcessor.java:271]: How should works this line?
`MyAbstractImplementation` looks like usual non-abstract class..

**Review by DarkWeird — CHANGES_REQUESTED** [review]
This PR breaking getting beans by interfaces and abstracts.
1. `CollectionResolveTest` is broken
2.  modified copy of `AbstractBeanTest#injectAbstractBean`:
```

    @Test
    public void injectAbstractBeanGetAbstract() {
        ServiceRegistry registry = new ServiceRegistry();
        registry.with(AbstractBeanTest.Impl1.class).lifetime(Lifetime.Singleton);
        registry.with(AbstractBeanTest.Impl2.class).lifetime(Lifetime.Singleton);
        registry.with(ConcreteImplementationFromAbs.class).lifetime(Lifetime.Singleton);

        BeanContext cntx = new DefaultBeanContext(registry);
        MyAbstractImplementation c1 = cntx.getBean(MyAbstractImplementation.class);

        Assert.assertNotNull(c1.impl);

    }
```
this modified test for getting as super class. don't works.

File: gestalt-inject-java/src/main/java/org/terasology/gestalt/annotation/processing/BeanDefinitionProcessor.java (hunk 2 above)
Comment (pollend) [reply gestalt-inject-java/src/main/java/org/terasology/gestalt/annotation/processing/BeanDefinitionProcessor.java:271]: some classes can't be instantiated. I guess the logic is faulty. :/

**Review by DarkWeird — APPROVED** [review]

**Pull request merged**